textfsm/terminal.py

```diff
@@ -260,57 +313,96 @@ class Pager(object):
-    try:
-      self._tty = open('/dev/tty')
-    except IOError:
-      # No TTY, revert to stdin
-      self._tty = sys.stdin
-    self.SetLines(None)
+    self._text = text
     self.Reset()
+    self.SetLines()
+    # Add 0.005 sec delay between lines.
+    if delay:
+      self._delay = 0.005
+    else:
+      self._delay = 0
 
-  def __del__(self):
-    """Deconstructor, closes tty."""
-    if getattr(self, '_tty', sys.stdin) is not sys.stdin:
-      self._tty.close()
-
-  def Reset(self):
+  def Reset(self) -> None:
     """Reset the pager to the top of the text."""
-    self._displayed = 0
-    self._currentpagelines = 0
-    self._lastscroll = 1
-    self._lines_to_show = self._cli_lines
+    self.first_line = 0
 
-  def SetLines(self, lines):
-    """Set number of screen lines.
+  def SetLines(self, num_lines: int = 0) -> typing.Tuple[int, int]:
+    """Set number of lines to display at a time.
 
     Args:
-      lines: An int, number of lines. If None, use terminal dimensions.
+      num_lines: An int, number of lines. If 0 use terminal dimensions.
+        Maximum number should be one less than full terminal height,
+        to allow for a user prompt. 
 
     Raises:
       ValueError, TypeError: Not a valid integer representation.
-    """
 
-    (self._cli_cols, self._cli_lines) = shutil.get_terminal_size()
-
-    if lines:
-      self._cli_lines = int(lines)
+    Returns:
+      Tuple, the width and lines of the terminal.
+    """
 
-  def Clear(self):
+    # Get the terminal size.
+    (cols, lines) = shutil.get_terminal_size()
+    # If we want paging by other than a whole window height.
+    # For a whole window height, we drop one line to leave room for prompting.
+    self._lines = int(num_lines) or lines - 1
+    # Must be at least two rows, one row of output and one for the prompt.
+    self._lines = max(2, self._lines)
+    # Only number of rows is user configurable, we keep the terminal width.
+    self._cols = cols
+    return (self._cols, self._lines)
+
+  def Clear(self) -> None:
     """Clear the text and reset the pager."""
     self._text = ''
     self.Reset()
 
-  def Page(self, text=None, show_percent=None):
+  def _Display(self, start: int, length: int = 0
+               ) -> typing.Tuple[int, float, int]:
+    """Display a range of lines from the text.
+
+    Args:
+      start: An int, the first line to display.
+      length: An int, the number of lines to display.
+    Returns:
+      Tuple, the next line after, and a percentage for where that line is.
+    """
+
+    # Break text on newlines. But also break on line wrap.
+    text_list = LineWrap(self._text).splitlines()
+    total_length = len(text_list)
+
+    # Bound start and end to be within the text.
+    start = max(0, start)
+    # If open-ended, trim to be whole of text.
+    if not length:
+      end = total_length
+    else:
+      end = min(start + length, total_length)
+
+    # Clear the screen.
+    self._WriteOut('\033[2J\033[H')
```

Comment:
should this be a const at the top like other sequences?

Reply:
Done